### PR TITLE
Adsk contrib - Add support for neon intrinsic integration

### PR DESCRIPTION
**Pull request by cedrik-fuoco-adsk**
This is a PR that is replacing the PR about ARM Neon here: https://github.com/AcademySoftwareFoundation/OpenColorIO/pull/1775.

The new PR contains all the previous code changes that were already accepted in the previous PR and contains new commits that were needed to integrate it with the code from the PR [Add AVX2/AVX/SSE2 SIMD accelerated 1D/3D LUTS](https://github.com/AcademySoftwareFoundation/OpenColorIO/pull/1687) (merged in main already).

Therefore, please ignore the first commit named "Merging the previous ARM Neon branch".

### CODE REVIEW

File: src/OpenColorIO/CPUInfo.cpp [hunk 1]
```diff
@@ -182,6 +182,46 @@ CPUInfo& CPUInfo::instance()
...
+
+    // Also note that during testing on a Apple M1 MacBook, SSE 4.2 does not seems to be supported.
+
+#if !defined(__aarch64__)
+    // Enable SSE2 instructions support using Rosetta for the x86_64 architecture on Apple ARM cpu.
+    if (OCIO_USE_SSE2)
+    {
+        flags |= X86_CPU_FLAG_SSE2;
+    }
+    //TODO: Once the other SSE instructions are implemented into OCIO, these can be enabled here.
```
Comment (markreidvfx) [inline src/OpenColorIO/CPUInfo.cpp:205]: why do we need to special case x86 for rosetta? Rosetta supports the `cpuid` instruction, can't we just use the regular path?

My M2 mac reports these extensions 
x86_64 VirtualApple @ 2.50GHz +sse2+sse3+ssse3+sse4+sse42

Comment (cedrik-fuoco-adsk) [reply src/OpenColorIO/CPUInfo.cpp:205]: We don't. I came to the same conclusion after looking at the failed CIs.

The next commit should fix the CI and Rosetta will use the regular path for x86.

**Comment by remia** [comment]
It seems this PR fixes the universal build from Intel based macOS, so it could fix the failing Python wheels nightly build, will be good to check, once this is merged.

File: share/cmake/utils/CheckSupportAVX512.cmake [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+file(WRITE "${CMAKE_BINARY_DIR}/CMakeTmp/avx512_test.cpp" "${AVX512_CODE}")
+
+message(STATUS "Performing Test COMPILER_SUPPORTS_AVX512")
+try_compile(COMPILER_SUPPORTS_AVX512
+  "${CMAKE_BINARY_DIR}/CMakeTmp"
+  "${CMAKE_BINARY_DIR}/CMakeTmp/avx512_test.cpp"
+)
+
+if(COMPILER_SUPPORTS_AVX512)
+  message(STATUS "Performing Test COMPILER_SUPPORTS_AVX512 - Success")
```
Comment (michdolan) [inline share/cmake/utils/CheckSupportAVX512.cmake:44]: [very minor] Three of these lines are under-indented (2 instead of 4 spaces) compared to the rest of the module. This is present in all of the `CheckSupport*.cmake` modules. Could be cleaned up later, but just mentioning it since I am reading through everything.

File: src/OpenColorIO/SSE.h [hunk 3]
```diff
@@ -20,6 +27,34 @@
...
+#if defined(__aarch64__)
+    #if defined(OCIO_USE_SSE2NEON)
+        // Using vmaxnmq_f32 and vminnmq_f32 rather than sse2neon's vmaxq_f32 and vminq_f32 due to 
+        // NaN handling. This doesn't seem to be significantly slower than the default sse2neon behavior.
+
+        // With the Intel intrinsics, if one value is a NaN, the second argument is output, as if it were 
+        // a simple (a>b) ? a:b. OCIO sometimes uses this behavior to filter out a possible NaN in the 
+        // first argument. The vmaxq/vminq will return a NaN if either input is a NaN, which omits the 
+        // filtering behavior. The vmaxnmq/vminnmq (similar to std::fmax/fmin) are not quite the same as 
+        // the Intel _mm_max_ps / _mm_min_ps since they always returns the non-NaN argument
```
Comment (michdolan) [inline src/OpenColorIO/SSE.h:42]:
```suggestion
        // the Intel _mm_max_ps / _mm_min_ps since they always return the non-NaN argument 
```

File: src/OpenColorIO/SSE2.h [hunk 4]
```diff
@@ -20,6 +29,33 @@
...
+#if defined(__aarch64__)
+    #if defined(OCIO_USE_SSE2NEON)
+        // Using vmaxnmq_f32 and vminnmq_f32 rather than sse2neon's vmaxq_f32 and vminq_f32 due to 
+        // NaN handling. This doesn't seem to be significantly slower than the default sse2neon behavior.
+
+        // With the Intel intrinsics, if one value is a NaN, the second argument is output, as if it were 
+        // a simple (a>b) ? a:b. OCIO sometimes uses this behavior to filter out a possible NaN in the 
+        // first argument. The vmaxq/vminq will return a NaN if either input is a NaN, which omits the 
+        // filtering behavior. The vmaxnmq/vminnmq (similar to std::fmax/fmin) are not quite the same as 
+        // the Intel _mm_max_ps / _mm_min_ps since they always returns the non-NaN argument
```
Comment (michdolan) [inline src/OpenColorIO/SSE2.h:44]:
```suggestion
        // the Intel _mm_max_ps / _mm_min_ps since they always return the non-NaN argument 
```

**Review by michdolan — APPROVED** [review]
Added a couple really minor notes which could be cleaned up later.

**Comment by markreidvfx** [comment]
Turning off SSE2NEON with  `-DOCIO_USE_SSE2NEON=OFF` fails to compile. Looks like clang is trying to compile SSE code.
```
In file included from /Users/mark/Dev/OpenColorIO/src/OpenColorIO/ops/cdl/CDLOpCPU.cpp:12:
/Users/mark/Dev/OpenColorIO/src/OpenColorIO/SSE.h:45:23: error: unknown type name '__m128'
        static inline __m128 _mm_max_ps(__m128 a, __m128 b)
```

On apple arm64 i'm getting 5 test_cpu_exec failed tests 
```
[ 153/1049] [CPUProcessor / optimizations                                ] - FAILED
[ 417/1049] [FileFormatCTF / bake_1d_3d                                  ] - FAILED
[ 427/1049] [FileFormatHDL / bake_1d_shaper                              ] - FAILED
[ 462/1049] [FileFormatResolveCube / bake_1d_shaper                      ] - FAILED
[ 471/1049] [FileFormatSpi1D / bake_1d_shaper                            ] - FAILED
```

I don't think they are related to this pull request as I get the same ones from the current main branch. I haven't dug into it but they look kind of like round errors.

All the cpu test pass under rosetta.

File: share/cmake/utils/CheckSupportAVX.cmake [hunk 5]
```diff
@@ -0,0 +1,60 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright Contributors to the OpenColorIO Project.
+
+set(_cmake_cxx_flags_orig "${CMAKE_CXX_FLAGS}")
+
+if(APPLE AND "${CMAKE_OSX_ARCHITECTURES}" MATCHES "arm64;x86_64"
```
Comment (remia) [inline share/cmake/utils/CheckSupportAVX.cmake:6]: Add parenthesis after the `AND` to make it clearer (same applies if the other check* modules)?

File: share/cmake/utils/CheckSupportF16C.cmake [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    #include <immintrin.h>
+
+    int main() 
+    { 
+        _mm_cvtph_ps(_mm_set1_epi16(0x3C00)); 
+        return 0; 
+    }
+")
+
+file(WRITE "${CMAKE_BINARY_DIR}/CMakeTmp/sse42_test.cpp" "${F16C_CODE}")
```
Comment (remia) [inline share/cmake/utils/CheckSupportF16C.cmake:35]: Very minor: seems like the temp file name was not updated.

File: share/cmake/modules/install/Installsse2neon.cmake [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+# Using FetchContent_GetProperties and FetchContent_Populate instead.
+FetchContent_GetProperties(sse2neon)
+
+if(NOT sse2neon_POPULATED)
+  FetchContent_Populate(sse2neon)
+
+  set(_EXT_DIST_INCLUDE "${CMAKE_BINARY_DIR}/ext/dist/${CMAKE_INSTALL_INCLUDEDIR}")
+  file(COPY "${sse2neon_SOURCE_DIR}/sse2neon.h" DESTINATION "${_EXT_DIST_INCLUDE}/sse2neon")
+
+  # sse2neon_INCLUDE_DIR is used internally for CheckSupportSSE2.cmake and to create sse2neon
```
Comment (remia) [inline share/cmake/modules/install/Installsse2neon.cmake:32]: Very minor: shouldn't it say CheckSupportSSEUsingSSE2NEON.cmake instead of CheckSupportSSE2.cmake?

File: docs/quick_start/installation.rst [hunk 8]
```diff
@@ -277,14 +278,37 @@ Here are the most common OCIO-specific CMake options (the default values are sho
...
+
+On the MacOS, the default is to build universal binaries 
+(natively supporting both the Intel and ARM processors). The ``-DCMAKE_OSX_ARCHITECTURES`` option 
+may be set to just arm64 or x86_64 to override the default value, which is ``arm64;x86_64``.
+
+When doing a universal build, note that the OCIO dependencies must be built as universal libraries 
+too. If you are running in OCIO_INSTALL_EXT_PACKAGES=MISSING or NONE mode, your build will fail if 
+any of your installed libraries are not universal. The easiest way to address this is to set 
+OCIO_INSTALL_EXT_PACKAGES=ALL in order to let OCIO build everything. Alternatively, you may set 
+CMAKE_OSX_ARCHITECTURES to just the platform you are targeting.
```
[the remark this answers is not in the document]
Comment (remia) [reply docs/quick_start/installation.rst:310]: Trying to look at a few other ASWF projects (Imath, OpenEXR, MaterialX, OIIO), I don't see any using universal build as the default setting I think (they do not seem to touch `CMAKE_OSX_ARCHITECTURES`). I wonder if there are other projects doing this?

[the remark this answers is not in the document]
Comment (cedrik-fuoco-adsk) [reply docs/quick_start/installation.rst:310]: I agree with your concerns. It might not be the best default if we consider an older MacBook that won't be able to build for ARM. 

However, I assume that the most recent MacBook (intel or arm) can build the universal build?

**Comment by cedrik-fuoco-adsk** [comment]
Thanks @remia, @markreidvfx  and @michdolan for your comments. 

Everything should have been covered by the new commits.

> do we need to add / update the macOS build variants in CI to cover all the new configurations?

I think that would be a good idea. I'll add a test for OCIO_USE_SSE2NEON at the minimum

File: docs/quick_start/installation.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (cedrik-fuoco-adsk) [reply docs/quick_start/installation.rst:310]: I will be removing the universal build by default.

[the remark this answers is not in the document]
Comment (remia) [reply docs/quick_start/installation.rst:310]: Yeah I think my concerns are mostly that the builds will most likely not work anymore for most users unless using `OCIO_INSTALL_EXT_PACKAGES=ALL` and also will take more time, resources and size than needed for people that are not packaging OCIO but only doing local builds. I guess also another thing would be to align with other projects using CMake to have predictable behaviour, not sure if there is a consensus here, I only sampled 4 projets.

**Comment by remia** [comment]
@cedrik-fuoco-adsk I think only 20 jobs can run in parallel on a workflow but nothing prevent us from having more than 20, it will just take longer. Another possibility is also to add new variants to a nightly build if these are rare build settings.

File: docs/quick_start/installation.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (cedrik-fuoco-adsk) [reply docs/quick_start/installation.rst:310]: I think it makes sense to keep the current architecture as the default build. I'll remove the code from the CMakeLists.txt

File: src/OpenColorIO/CPUInfoConfig.h.in [hunk 9]
```diff
@@ -2,14 +2,50 @@
...
+// aarch64 defines.
+
+// Building for x86_64 processor on a non-ARM host architecture
+// OR Building on/for an ARM architecture and using SSE2NEON.
+#if (OCIO_ARCH_X86 && !defined(__aarch64__)) || (defined(__aarch64__) && OCIO_USE_SSE2NEON)
+    #cmakedefine01 OCIO_USE_SSE2
+    #cmakedefine01 OCIO_USE_SSE3
+    #cmakedefine01 OCIO_USE_SSSE3
+    #cmakedefine01 OCIO_USE_SSE4
+    #cmakedefine01 OCIO_USE_SSE42
```
Comment (markreidvfx) [inline src/OpenColorIO/CPUInfoConfig.h.in:28]: Building non universal arm64, I still needed to specify `-DOCIO_USE_SSE2=ON` etc to turn all these on. `-DOCIO_USE_SSE2NEON=ON` should probably imply all these SSE features are ON.

Comment (markreidvfx) [reply src/OpenColorIO/CPUInfoConfig.h.in:28]: They are not turning on for me when I build natively on arm64. I've try cleaning my env several times.
I'm just doing the basic `mkdir build && cd build && cmake .. && cmake --build .`

and my `CPUInfoConfig.h` shows

```
#if (OCIO_ARCH_X86 && !defined(__aarch64__)) || (defined(__aarch64__) && OCIO_USE_SSE2NEON)
    #define OCIO_USE_SSE2 0
    #define OCIO_USE_SSE3 0
    #define OCIO_USE_SSSE3 0
    #define OCIO_USE_SSE4 0
    #define OCIO_USE_SSE42 0
```

It looks like its because `CMAKE_OSX_ARCHITECTURES` isn't set to anything when I build it this way.
That section in [CMakeLists.txt](https://github.com/AcademySoftwareFoundation/OpenColorIO/pull/1828/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR192-R202) need it to be `arm64`.

`(APPLE AND "${CMAKE_SYSTEM_PROCESSOR}" MATCHES "arm64" AND "${CMAKE_OSX_ARCHITECTURES}" MATCHES "arm64")`

`cmake ..  -DCMAKE_OSX_ARCHITECTURES=arm64` works.

Should we set CMAKE_OSX_ARCHITECTURES = CMAKE_SYSTEM_PROCESSOR if undefined?

Comment (cedrik-fuoco-adsk) [reply src/OpenColorIO/CPUInfoConfig.h.in:28]: Oh right, I get it now. There's a missing condition in the CMakeLists.txt. I'll update it.

Comment (cedrik-fuoco-adsk) [reply src/OpenColorIO/CPUInfoConfig.h.in:28]: The latest commit should have fixed the issue (I've tested it on my M1). @markreidvfx

Comment (markreidvfx) [reply src/OpenColorIO/CPUInfoConfig.h.in:28]: That fixes it for me on my M2 mac too.

File: docs/quick_start/installation.rst [hunk 10]
```diff
@@ -278,14 +279,31 @@ Here are the most common OCIO-specific CMake options (the default values are sho
...
 - ``-DOCIO_BUILD_DOCS=OFF`` (Set to ON to build the documentation)
 - ``-DOCIO_BUILD_FROZEN_DOCS=OFF`` (Set to ON to update the Python documentation)
 
+Note that OCIO will turn off any specific SIMD CPU performance optimizations if they are not supported 
+by the build target architecture. The default for ``OCIO_USE_SSE2``, ``OCIO_USE_AVX``, ``OCIO_USE_AVX2`` and 
+``OCIO_USE_F16C`` depends on the architecture, but will be ON where supported.
+
+On MacOS, the default is to build for the native architecture. The ``-DCMAKE_OSX_ARCHITECTURES`` option 
+may be set to ``arm64;x86_64`` to build the universal binaries.
+
```
Comment (markreidvfx) [inline docs/quick_start/installation.rst:300]: I double checked, the order the architectures doesn't cause any issues, but we might want to reword this to show you need quotes around the architectures `-DCMAKE_OSX_ARCHITECTURES"arm64;x86_64"`. It took me like 10mins to realize the semicolon `;` in bash or zshell ends a command :p

**Comment by doug-walker** [comment]
Thanks for your help with this @markreidvfx !

We updated the install instructions accordingly.  Going to merge this as the expected tests passed before we updated the documentation.
